psdisk/psdisk.c

```diff
@@ -117,199 +116,220 @@ static void psdisk_showPartsTable(void)
-		return -1;
+	if (fread(&count, 1, sizeof(count), psdisk_common.file) != sizeof(count)) {
+		return NULL;
 	}
 
-	return fread(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
-}
+#if __BYTE_ORDER != __LITTLE_ENDIAN
+	count = le32toh(count);
+#endif
 
+	size = ptable_size(count);
+	if (size > psdisk_common.blksz) {
+		return NULL;
+	}
 
-static ssize_t psdisk_writeImg(unsigned int addr, const void *buff, size_t bufflen)
-{
-	ssize_t res;
-	char *wBuff;
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return NULL;
+	}
 
-	if (bufflen > psdisk_common.mem.sectorSize) {
-		fprintf(stderr,"Partition table exceeds sector size. Reduce number of partitions or increase sector size.");
-		return -1;
+	ptable = malloc(size);
+	if (ptable == NULL) {
+		return NULL;
 	}
 
-	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
-		fprintf(stderr,"Cannot change file position, err: %s.\n", strerror(errno));
-		return -1;
+	if (fread(ptable, 1, size, psdisk_common.file) != size) {
+		free(ptable);
+		return NULL;
 	}
 
-	if ((wBuff = (char *)malloc(psdisk_common.mem.sectorSize)) == NULL) {
-		fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
-		return -1;
+	if (ptable_deserialize(ptable, psdisk_common.memsz, psdisk_common.blksz) < 0) {
+		free(ptable);
+		return NULL;
 	}
 
-	memset(wBuff, 0xff, psdisk_common.mem.sectorSize);
+	return ptable;
+}
 
-	if (PTABLE_WITH_OFFSET(psdisk_common.opts)) {
-		while (addr > 0) {
-			fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
-			addr -= psdisk_common.mem.sectorSize;
-		}
+
+static int psdisk_writeImg(ptable_t *ptable)
+{
+	uint32_t size;
+
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		return -1;
 	}
+	size = ptable_size(ptable->count);
 
-	memcpy(wBuff, buff, bufflen);
-	res = fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
+	if (ptable_serialize(ptable, psdisk_common.memsz, psdisk_common.blksz) < 0) {
+		return -1;
+	}
 
-	free(wBuff);
+	if (fwrite(ptable, 1, size, psdisk_common.file) != size) {
+		return -1;
+	}
 
-	return res;
+	return 0;
 }
 
 
 /* Partition table modification functions */
 
 static int psdisk_readPartsTable(void)
 {
-	int i;
-	uint32_t pCnt;
-	struct plist_node_t *pNode;
-	ptable_partition_t *pHeaders;
+	struct plist_node_t *node;
+	ptable_t *ptable;
+	uint32_t i;
+	int ret = 0;
 
-	if ((pHeaders = ptable_readPartitions(&pCnt, &psdisk_common.mem)) == NULL) {
-		fprintf(stderr,"The file contains incorrect partition table.\n");
+	ptable = psdisk_readImg();
+	if (ptable == NULL) {
+		fprintf(stderr, "The file contains incorrect partition table.\n");
 		return -1;
 	}
 
-	for (i = 0; i < pCnt; ++i) {
-		if ((pNode = (struct plist_node_t *)calloc(1, sizeof(struct plist_node_t))) == NULL) {
-			fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
-			free(pHeaders);
-			return -1;
+	for (i = 0; i < ptable->count; i++) {
+		node = calloc(1, sizeof(struct plist_node_t));
```

Comment:
Somewhat more error-resistant pattern
```suggestion
		node = calloc(1, sizeof(*node));
```

Reply:
Fixed all occurrences.